flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/plan/StreamOptimizer.scala

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.plan
+
+import org.apache.calcite.plan.hep.HepMatchOrder
+import org.apache.calcite.rel.RelNode
+import org.apache.calcite.sql2rel.RelDecorrelator
+import org.apache.calcite.tools.{RelBuilder, RuleSet, RuleSets}
+import org.apache.flink.table.calcite.{CalciteConfig, RelTimeIndicatorConverter}
+import org.apache.flink.table.plan.nodes.FlinkConventions
+import org.apache.flink.table.plan.nodes.datastream.UpdateAsRetractionTrait
+import org.apache.flink.table.plan.rules.FlinkRuleSets
+import org.apache.flink.table.planner.PlanningConfigurationBuilder
+
+import scala.collection.JavaConverters._
+
+/**
+  * An [[Optimizer]] that can be used for optimizing a streaming plan.
+  *
+  * @param calciteConfig provider for [[CalciteConfig]]. It is a provider because the
+  *                      [[org.apache.flink.table.api.TableConfig]] in a
+  *                      [[org.apache.flink.table.api.TableEnvImpl]] is mutable.
+  * @param planningConfigurationBuilder provider for [[org.apache.calcite.plan.RelOptPlanner]] and
+  *                                     [[org.apache.calcite.plan.Context]]
+  */
+class StreamOptimizer(
+    calciteConfig: () => CalciteConfig,
+    planningConfigurationBuilder: PlanningConfigurationBuilder)
+  extends Optimizer(calciteConfig, planningConfigurationBuilder) {
+
+  /**
+    * Generates the optimized [[RelNode]] tree from the original relational node tree.
+    *
+    * @param relNode The root node of the relational expression tree.
+    * @param updatesAsRetraction True if the sink requests updates as retraction messages.
+    * @return The optimized [[RelNode]] tree
+    */
+  def optimize(
+    relNode: RelNode,
+    updatesAsRetraction: Boolean,
+    relBuilder: RelBuilder): RelNode = {
+    val convSubQueryPlan = optimizeConvertSubQueries(relNode)
+    val expandedPlan = optimizeExpandPlan(convSubQueryPlan)
+    val decorPlan = RelDecorrelator.decorrelateQuery(expandedPlan, relBuilder)
+    val planWithMaterializedTimeAttributes =
+      RelTimeIndicatorConverter.convert(decorPlan, relBuilder.getRexBuilder)
```

Comment:
Just something I noticed: documentation is a bit sparse and it's hard for me to find out what a `RelTimeIndicatorConverter` actually does, because it has no class-level (or Object-level) documentation. 😅 

Not for this PR to fix, though...

Reply:
Totally agree. We try to add proper javadocs for the newly introduced interfaces/classes.